packages/kit/src/core/dev/index.js

```diff
@@ -311,6 +274,47 @@ class Watcher extends EventEmitter {
 				}
 			});
 		});
+
+		/**
+		 * @type {vite.ViteDevServer}
+		 */
+		this.vite = await vite.createServer({
+			...user_config,
+			configFile: false,
+			root: this.cwd,
+			resolve: {
+				...user_config.resolve,
+				alias: {
+					...(user_config.resolve && user_config.resolve.alias),
+					$app: path.resolve(`${this.dir}/runtime/app`),
+					$lib: this.config.kit.files.lib
+				}
+			},
+			plugins: [
+				...(user_config.plugins || []),
+				svelte({
+					extensions: this.config.extensions,
+					emitCss: !this.config.kit.amp
+				})
+			],
+			publicDir: this.config.kit.files.assets,
+			server: {
+				...user_config.server,
+				middlewareMode: true,
+				hmr: {
+					...(user_config.server && user_config.server.hmr),
+					...(this.https ? { server: this.server, port: this.port } : {})
```

Comment:
oh, I see the issue now. I do think that `this.vite` should continue to be defined before `this.server` (sorry to somewhat contradict my last message. I wasn't totally following before)

what I don't understand here though is why you have a `server` attribute. that's not a valid Vite option as far as I can tell: https://vitejs.dev/config/#server-hmr. If you need to provide it. can you instead provide the `protocol`, `host`, and `port` without providing `this.server` in order to break the circular dependency?

Reply:
The hmr.server option is referenced in [Vite's web socket setup](https://github.com/vitejs/vite/blob/main/packages/vite/src/node/server/ws.ts).    The first few lines (23-35) is the applicable code.  It's true that it isn't in the documentation -- it may be an oversight, and perhaps we should verify that with the Vite team.

SvelteKit uses Vite's "middlewareMode" (e.g., where the server is managed by SvelteKit, and Vite is considered as middleware).  Thus, the websocket must use the SvelteKit's https server to ensure that the certificate is valid, i.e., the certificate is validated through a Certificate Authority, or the user opts to 'proceed' when using an unsigned certificate or otherwise suspicious certificate.

When connecting to a secure websocket from the browser, the browser automatically changes the request from 'wss' to 'https'.  If the certificate is valid, the connection will allowed.  Vite will then intercept this request and complete the wss connection.  The request is identified by the header "sec-websocket-protocol: vite-hmr".

'hmr.server' tells Vite which server will receive the connection request -- it needs the actual server object, as it's adding the 'upgrade' event handler to that object to complete the connection and to accept subsequent wss requests.

Regarding the dummy (interim) handler -- My concern is that some kind of error (possibly fatal?) would occur if the client requests the website before it's initialized.

In summary:

- We need to verify if the lack of documentation for the hmr.server option was an oversight, and that it will continue to be supported (how? who to contact?)
- SvelteKit's server object must be passed to hmr.server, as this is how Vite expects it.
- Should I revert the last commit?  Should the dummy (interim) handler be kept?  Should I wait?

Reply:
I discovered another work-around that doesn't require any code changes, but it doesn't support Android (Chrome, at least).  I don't know about IOS.  It's kind of ugly:

1. Add vite.server.https option to svelte.config.js:

```
	kit: {
		target: '#svelte',
		vite: {
			server: {
				https: true
			}
		}
	}
```
2. Start the dev server - npx svelte-kit dev -H --host 192.168.2.50   (or whichever ip points to your dev machine)
3. Point browser to https://192.168.2.50:24678
4. In the "Your connection is not private" dialog, click "Advanced" then "Proceed to ..."
5. The browser displays "upgrade required".  Ignore this.
6. Now point your browser to https://192.168.2.50:3000
7. In the "Your connection is not private" dialog, click "Advanced" then "Proceed to ..."
8. You're on your way!

You'll need to follow this procedure upon any server restart.  

Pros:   No code changes
Cons: A bit kludgy, doesn't support all devices.

Reply:
Hmm, yeah, none of these options seem great. I think it would be nice if the Vite folks could document how this should be done.

I looked through their issue tracker and found a recently resolved issue, which makes me wonder if this is possible now? Take a look at: https://github.com/vitejs/vite/pull/1992#issuecomment-843939223

Reply:
I looked at the comment -- they use DevilBox, which generates a Certificate Authority which you import into your browser.  It then generates signed certificates using the trusted CA.  I've created my own trusted CA in the past, and it's quite an involved process.  Additionally, some companies disallow adding Trusted CA's (e.g., via WIndows Group Policy).

I also tested with the latest release of Vite (2.3.4), which includes the changes that you mentioned.  Actually, that was the version I was testing with when I discovered the latest workaround.

I've reached out to the author of that PR [in this comment](https://github.com/vitejs/vite/pull/1992#issuecomment-850731007).  So, I'll wait for his response before proceeding further.

Reply:
I received this [this reply](https://github.com/vitejs/vite/pull/1992#issuecomment-850841408.)  It appears that this PR didn't specifically address self-signed certificates.
I opened [this issue](https://github.com/vitejs/vite/issues/3587) on Vite.  Hopefully someone will review it, and confirm that the option should be documented.

Reply:
I received confirmation from the Vite maintainers that the server.hmr.server option is an exposed (public) option.  They asked that I open a PR to update the documentation - https://github.com/vitejs/vite/pull/3590.  It's been approved and merged into Vite's main branch, awaiting the next patch release.

So, we now need to settle on an approach.  Whichever approach, we need to have SvelteKit's server created via `get_server()` before `vite.createServer()`, as the server handle must be passed to `vite.createServer()`.

Here are two approaches:

a. Switch the order of `this.vite = await vite.createServer()` block with `this.server = await get_server` block.  This was my latest commit: 50a05c2.

b. Move only the call to `get_server()` before the `this.vite = await vite.createServer()` block, and attach the handler `this.vite.middlewares(req, res, async () => {...}` in its current location.

For the latter option, I see a few ways of doing this:

1. Add a dummy handler, returning a 204 response (No Content) until the middlewares handler is attached.  This was my commit: b5779a9.
2. Add a dummy handler that retries every 'n' milliseconds until the middlewares handler is attached.
3. take the 'listen' out of `server/index.js` and issue the listen after the middlewares handler is attached.  This would also necessitate a change in the preview code (`start/index.js`)

My concern with (a) and (b.1) is: What happens when an http(s) request is received before everything is initialized?  Although rare, this could happen if a browser tab already points to the URL, and the vite client does its 'ping' during the server's uninitialized interval, or if the user is fast enough to reload the browser during this interval.

With option (b.1), the user would only need to refresh.  I'm not sure what would happen using option (a).

Personally, I like option (b.2), but I'd appreciate your unbiased opinion.

I will await your guidance before proceeding further.

Reply:
I would say that b.3. is definitely the best approach in my mind

Reply:
OK.  I committed the changes, but I had a problem running 'pnpm test' -- It succeeds, but, looking at the console log, I get this message:
```
(node:931474) UnhandledPromiseRejectionWarning: Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/home/jbusillo/Projects/kit-pull/packages/kit/test/apps/custom-extension/svelte.config.cjs' imported from /home/jbusillo/Projects/kit-pull/packages/kit/src/core/load_config/index.js
    at finalizeResolution (internal/modules/esm/resolve.js:276:11)
    at moduleResolve (internal/modules/esm/resolve.js:699:10)
    at Loader.defaultResolve [as _resolve] (internal/modules/esm/resolve.js:810:11)
    at Loader.resolve (internal/modules/esm/loader.js:88:40)
    at Loader.getModuleJob (internal/modules/esm/loader.js:241:28)
    at Loader.import (internal/modules/esm/loader.js:176:28)
    at importModuleDynamically (internal/modules/esm/translators.js:114:35)
    at exports.importModuleDynamicallyCallback (internal/process/esm_loader.js:30:14)
    at load_config (file:///home/jbusillo/Projects/kit-pull/packages/kit/src/core/load_config/index.js:109:60)
    at main (file:///home/jbusillo/Projects/kit-pull/packages/kit/test/test.js:376:24)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:931474) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:931474) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```
I see that there was a merge performed for #1595, which might be affecting this.  Is it possible that no one had noticed this unhighlighted console message while testing?  Or is my repository messed up?

Earlier today, I had done a "Fetch & Merge" on my pulled repository, since I had seen that Rich Harris had changed /dev/index.js after my initial PR, and wanted to include those changes in my repo.

Reply:
Hmmm. Weird. I'm getting this locally too. I don't know why we didn't see it on the CI

Reply:
Oh, nevermind. My error was because I hadn't run `pnpm build` before `pnpm test`. If I do that then the tests pass locally for me on `master`

Reply:
I created a new VSCode project by re-cloning from my pulled repository.  Tests are running fine.  I must have messed up my old project somehow.